File: src/main/scala/dynamok/DynamoDbSourceConnector.scala
package dynamok

import java.util.{List => JList, Map => JMap}

import com.amazonaws.regions.Regions
import com.amazonaws.services.dynamodbv2.model.{DescribeStreamRequest, DescribeStreamResult, Shard, StreamDescription, StreamViewType, TableDescription}
import com.amazonaws.services.dynamodbv2.{AmazonDynamoDBClient, AmazonDynamoDBStreamsClient}
import org.apache.kafka.common.config.ConfigDef
import org.apache.kafka.common.config.ConfigDef.{Importance, Type}
import org.apache.kafka.common.utils.AppInfoParser
import org.apache.kafka.connect.connector.Task
import org.apache.kafka.connect.source.SourceConnector
import org.apache.kafka.connect.util.ConnectorUtils
import org.slf4j.LoggerFactory

import scala.collection.JavaConverters._

class DynamoDbSourceConnector extends SourceConnector {

  import DynamoDbSourceConnector._

  val log = LoggerFactory.getLogger(classOf[DynamoDbSourceConnector])

  var region: Regions = _
  var tableDescription: TableDescription = _
  var shards: JList[Shard] = _
  var destTopic: String = _

  override def taskClass(): Class[_ <: Task] = classOf[DynamoDbSourceTask]

  override def taskConfigs(maxTasks: Int): JList[JMap[String, String]] = {
    ConnectorUtils.groupPartitions(shards, maxTasks).asScala.map {
      taskShards =>
        Map(
          ConfigKeys.Region -> region.getName,
          ConfigKeys.Table -> tableDescription.getTableName,
          ConfigKeys.Shards -> taskShards.asScala.map(_.getShardId).mkString(","),
          ConfigKeys.StreamArn -> tableDescription.getLatestStreamArn,
          ConfigKeys.Topic -> destTopic
        ).asJava
    }.toList.asJava
  }

  override def start(props: JMap[String, String]): Unit = {
    val config = ConnectorConfig(props)

    val client = new AmazonDynamoDBClient()
    client.configureRegion(config.region)

    tableDescription = client.describeTable(config.table).getTable
    if (!tableDescription.getStreamSpecification.isStreamEnabled) {
      sys.error("The DynamoDB table does not have streams enabled")
    }

    val streamViewType = tableDescription.getStreamSpecification.getStreamViewType
    if (!AcceptableStreamViewTypes.contains(streamViewType)) {
      sys.error(s"""The DynamoDB table's stream view type needs to be one of ${AcceptableStreamViewTypes.mkString("{,", ",", "}")}, was: ${streamViewType}""")
    }

    region = config.region

    val streamsClient: AmazonDynamoDBStreamsClient = new AmazonDynamoDBStreamsClient()
    streamsClient.configureRegion(region)

    val describeStreamResult: DescribeStreamResult =
      streamsClient.describeStream(new DescribeStreamRequest().withStreamArn(tableDescription.getLatestStreamArn))

    val streamDescription: StreamDescription = describeStreamResult.getStreamDescription
    shards = streamDescription.getShards

    destTopic = config.topic
  }

  override def stop(): Unit = {
  }

  override def config(): ConfigDef = ConfigDefinition

  override def version(): String = AppInfoParser.getVersion

}

object DynamoDbSourceConnector {

  val AcceptableStreamViewTypes = Seq(
    StreamViewType.NEW_IMAGE.name(),
    StreamViewType.NEW_AND_OLD_IMAGES.name()
  )

  object ConfigKeys {
    val Region = "region"
    val Table = "table"
    val Shards = "shards"
    val StreamArn = "stream_arn"
    val Topic = "topic"
  }

  val ConfigDefinition = new ConfigDef()
    .define(ConfigKeys.Region, Type.STRING, Importance.HIGH, "AWS region.")
    .define(ConfigKeys.Table, Type.STRING, Importance.HIGH, "Source DynamoDB table.")
    .define(ConfigKeys.Topic, Type.STRING, Importance.HIGH, "Destination Kafka topic.")

  case class ConnectorConfig(props: JMap[String, String]) {

    lazy val region: Regions = Regions.fromName(props.get(ConfigKeys.Region))

    lazy val table: String = props.get(ConfigKeys.Table)

    lazy val shards: Seq[String] = props.get(ConfigKeys.Shards).split(',').filterNot(_.isEmpty)

    lazy val streamArn: String = props.get(ConfigKeys.StreamArn)

    lazy val topic: String = props.get(ConfigKeys.Topic)

  }

}

File: src/main/java/dynamok/source/DynamoDbSourceConnector.java
/*
 * Copyright 2016 Shikhar Bhushan
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package dynamok.source;

import com.amazonaws.auth.BasicAWSCredentials;
import com.amazonaws.auth.DefaultAWSCredentialsProviderChain;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDBStreamsClient;
import com.amazonaws.services.dynamodbv2.model.DescribeStreamRequest;
import com.amazonaws.services.dynamodbv2.model.DescribeStreamResult;
import com.amazonaws.services.dynamodbv2.model.ListTablesResult;
import com.amazonaws.services.dynamodbv2.model.Shard;
import com.amazonaws.services.dynamodbv2.model.StreamSpecification;
import com.amazonaws.services.dynamodbv2.model.StreamViewType;
import com.amazonaws.services.dynamodbv2.model.TableDescription;
import dynamok.Version;
import org.apache.kafka.common.config.ConfigDef;
import org.apache.kafka.connect.connector.Task;
import org.apache.kafka.connect.errors.ConnectException;
import org.apache.kafka.connect.source.SourceConnector;
import org.apache.kafka.connect.util.ConnectorUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

public class DynamoDbSourceConnector extends SourceConnector {

    private final Logger log = LoggerFactory.getLogger(getClass());

    private ConnectorConfig config;
    private Map<Shard, TableDescription> streamShards;

    @Override
    public Class<? extends Task> taskClass() {
        return DynamoDbSourceTask.class;
    }

    @Override
    public List<Map<String, String>> taskConfigs(int maxTasks) {
        return ConnectorUtils.groupPartitions(new ArrayList<>(streamShards.keySet()), maxTasks).stream().map(taskShards -> {
            final Map<String, String> taskConfig = new HashMap<>();
            taskConfig.put(TaskConfig.Keys.REGION, config.region.getName());
            taskConfig.put(TaskConfig.Keys.TOPIC_FORMAT, config.topicFormat);
            taskConfig.put(TaskConfig.Keys.SHARDS, taskShards.stream().map(Shard::getShardId).collect(Collectors.joining(",")));
            taskShards.forEach(shard -> {
                final TableDescription tableDesc = streamShards.get(shard);
                taskConfig.put(shard.getShardId() + "." + TaskConfig.Keys.TABLE, tableDesc.getTableName());
                taskConfig.put(shard.getShardId() + "." + TaskConfig.Keys.STREAM_ARN, tableDesc.getLatestStreamArn());
            });
            taskConfig.put(TaskConfig.Keys.LOG_RECORD_STREAM, config.isRecordStreamLogOn);
            return taskConfig;
        }).collect(Collectors.toList());
    }

    @Override
    public void start(Map<String, String> props) {
        config = new ConnectorConfig(props);
        streamShards = new HashMap<>();

        final AmazonDynamoDBClient client;
        final AmazonDynamoDBStreamsClient streamsClient;

        if (config.accessKeyId.value().isEmpty() || config.secretKey.value().isEmpty()) {
            client = new AmazonDynamoDBClient(DefaultAWSCredentialsProviderChain.getInstance());
            streamsClient = new AmazonDynamoDBStreamsClient(DefaultAWSCredentialsProviderChain.getInstance());
            log.debug("AmazonDynamoDBStreamsClient created with DefaultAWSCredentialsProviderChain");
        } else {
            final BasicAWSCredentials awsCreds = new BasicAWSCredentials(config.accessKeyId.value(), config.secretKey.value());
            client = new AmazonDynamoDBClient(awsCreds);
            streamsClient = new AmazonDynamoDBStreamsClient(awsCreds);
            log.debug("AmazonDynamoDB clients created with AWS credentials from connector configuration");
        }

        client.configureRegion(config.region);
        streamsClient.configureRegion(config.region);

        final Set<String> ignoredTables = new HashSet<>();
        final Set<String> consumeTables = new HashSet<>();

        String lastEvaluatedTableName = null;
        do {
            final ListTablesResult listResult = client.listTables(lastEvaluatedTableName);

            for (String tableName : listResult.getTableNames()) {
                if (!acceptTable(tableName)) {
                    ignoredTables.add(tableName);
                    continue;
                }

                final TableDescription tableDesc = client.describeTable(tableName).getTable();

                final StreamSpecification streamSpec = tableDesc.getStreamSpecification();

                if (streamSpec == null || !streamSpec.isStreamEnabled()) {
                    throw new ConnectException(String.format("DynamoDB table `%s` does not have streams enabled", tableName));
                }

                final String streamViewType = streamSpec.getStreamViewType();
                if (!streamViewType.equals(StreamViewType.NEW_IMAGE.name()) && !streamViewType.equals(StreamViewType.NEW_AND_OLD_IMAGES.name())) {
                    throw new ConnectException(String.format("DynamoDB stream view type for table `%s` is %s", tableName, streamViewType));
                }

                final DescribeStreamResult describeStreamResult =
                        streamsClient.describeStream(new DescribeStreamRequest().withStreamArn(tableDesc.getLatestStreamArn()));

                for (Shard shard : describeStreamResult.getStreamDescription().getShards()) {
                    streamShards.put(shard, tableDesc);
                }

                consumeTables.add(tableName);
            }

            lastEvaluatedTableName = listResult.getLastEvaluatedTableName();
        } while (lastEvaluatedTableName != null);

        log.info("Tables to ignore: {}", ignoredTables);
        log.info("Tables to ingest: {}", consumeTables);

        client.shutdown();
        streamsClient.shutdown();
    }

    private boolean acceptTable(String tableName) {
        return tableName.startsWith(config.tablesPrefix)
                && (config.tablesWhitelist.isEmpty() || config.tablesWhitelist.contains(tableName))
                && !config.tablesBlacklist.contains(tableName);
    }

    @Override
    public void stop() {
    }

    @Override
    public ConfigDef config() {
        return ConnectorConfig.CONFIG_DEF;
    }

    @Override
    public String version() {
        return Version.get();
    }

}
